measureme/src/raw_event.rs

```diff
@@ -1,46 +1,160 @@
+
+        let start_time_lower = start_nanos as u32;
+        let end_time_lower = end_nanos as u32;
+
+        let start_time_upper = (start_nanos >> 16) as u32 & 0xFFFF_0000;
+        let end_time_upper = (end_nanos >> 32) as u32;
+
+        let start_and_end_upper = start_time_upper | end_time_upper;
+
         RawEvent {
             event_kind,
             event_id,
             thread_id,
-            start_ns,
-            end_ns,
+            start_time_lower,
+            end_time_lower,
+            start_and_end_upper,
         }
     }
 
     #[inline]
     pub fn new_instant(
         event_kind: StringId,
         event_id: StringId,
-        thread_id: u64,
+        thread_id: u32,
         timestamp_ns: u64,
     ) -> RawEvent {
+        assert!(timestamp_ns <= MAX_INSTANT_TIMESTAMP);
+
+        let start_time_lower = timestamp_ns as u32;
+        let end_time_lower = 0xFFFF_FFFF;
+
+        let start_time_upper = (timestamp_ns >> 16) as u32;
+        let start_and_end_upper = start_time_upper | 0x0000_FFFF;
+
         RawEvent {
             event_kind,
             event_id,
             thread_id,
-            start_ns: timestamp_ns,
-            end_ns: std::u64::MAX,
+            start_time_lower,
+            end_time_lower,
+            start_and_end_upper,
+        }
+    }
+
+    #[inline]
+    pub fn start_nanos(&self) -> u64 {
+        self.start_time_lower as u64 | (((self.start_and_end_upper & 0xFFFF_0000) as u64) << 16)
+    }
+
+    #[inline]
+    pub fn end_nanos(&self) -> u64 {
+        self.end_time_lower as u64 | (((self.start_and_end_upper & 0x0000_FFFF) as u64) << 32)
+    }
+
+    #[inline]
+    pub fn is_instant(&self) -> bool {
+        self.end_nanos() == INSTANT_TIMESTAMP_MARKER
+    }
+
+    #[inline]
+    pub fn serialize(&self, bytes: &mut [u8]) {
+        assert!(bytes.len() == std::mem::size_of::<RawEvent>());
+
+        #[cfg(target_endian = "little")]
+        {
+            let raw_event_bytes: &[u8] = unsafe {
+                std::slice::from_raw_parts(
+                    self as *const _ as *const u8,
+                    std::mem::size_of::<RawEvent>(),
+                )
+            };
+
+            bytes.copy_from_slice(raw_event_bytes);
+        }
+
+        #[cfg(target_endian = "big")]
+        {
+            // We always emit data as little endian, which we have to do
+            // manually on big endian targets.
+            use byteorder::{LittleEndian, ByteOrder};
+
+            LittleEndian::write_u32(&mut bytes[0..], self.event_kind.as_u32());
+            LittleEndian::write_u32(&mut bytes[4..], self.event_id.as_u32());
+            LittleEndian::write_u32(&mut bytes[8..], self.thread_id);
+            LittleEndian::write_u32(&mut bytes[12..], self.start_time_lower);
+            LittleEndian::write_u32(&mut bytes[16..], self.end_time_lower);
+            LittleEndian::write_u32(&mut bytes[20..], self.start_and_end_upper);
+        }
```

Comment:
Out of curiosity, was the "manual" version benchmarked on x64?
The only reason I can think of is that it might be slower to write six `u32`s, one at a time (as opposed to 3 `u64` or one `u128` and one `u64`), but that's possible to alleviate, but first serializing to larger integers and then writing them out.

Reply:
IIRC the two versions were almost the same with the manual one having one instruction more (and the whole fn being something 10 instructions). I don't expect there to be much difference in practice since there is more expensive stuff going (like writing this data to disk).